pyuvdata/uvbeam.py

```diff
@@ -1084,6 +1084,60 @@ def __iadd__(self, other):
         self.__add__(other, inplace=True)
         return self
 
+    def _stokes_matrix(self, n):
+        """
+        Calculates Pauli matrices (where indices are reordered from the quantum mechanical convention
+        to an order which gives the ordering of the Stoes vector as [I, Q, U, V]) according to
+        https://arxiv.org/pdf/1401.2095.pdf.
+
+        Args :
+            n : Index for which to calculate the Pauli matrix; n should lie between 0 and 3 for the 4x4
+                 Mueller matrix in https://arxiv.org/pdf/1802.04151.pdf.
+
+        Returns :
+            2 x 2 numpy array Pauli matrix
+        """
+        if n < 0:
+            raise ValueError('n must be positive integer.')
+        if n > 4:
+            raise ValueError('n should lie between 0 and 3.')
+        if n == 0:
+            p = np.array([[1., 0.], [0., 1.]])
+        if n == 1:
+            p = np.array([[1., 0.], [0., -1.]])
+        if n == 2:
+            p = np.array([[0., 1.], [1., 0.]])
+        if n == 3:
+            p = np.array([[1., -1.j], [1.j, 0.]])
+
+        return p
+
+    def _construct_mueller(self, jones, i, j):
+        """
+        Generates Mueller component as done in https://arxiv.org/pdf/1802.04151.pdf
+
+                    Mij = Tr(J sigma_i J sigma_j J^*),
+
+        where sigma_i and sigma_j are Pauli matrices.
+
+        Args :
+            jones : Jones matrices containing the electric field components for the dipole arms or polarizations.
+
+            i : Index at which the first Pauli matrix sigma_i is calculated.
+
+            j : Index at which the second Pauli matrix sigma_j is calculated.
+
+        Returns :
+            npix numpy array containig the values for the corresponding Mueller component
+        """
+        Pi = self._stokes_matrix(i)
+        Pj = self._stokes_matrix(j)
+
+        Mij = 0.5 * np.einsum('...ab,...bc,...cd,...ad', Pi, jones, Pj, jones.conj())
+        Mij = np.abs(Mij)
+
+        return Mij
+
     def _get_beam(self, pol):
         """
         Get the healpix beam map corresponding to the specififed polarization,
```

Comment:
Why is this specific to Healpix?

Reply:
implemented in Kohn et al 2018, general case implementation n progress.